Providers/RustSyncManager.swift

```diff
@@ -0,0 +1,628 @@
+
+    public var lastSyncFinishTime: Timestamp? {
+        get {
+            return self.prefs.timestampForKey(PrefsKeys.KeyLastSyncFinishTime)
+        }
+
+        set(value) {
+            if let value = value {
+                self.prefs.setTimestamp(value,
+                                        forKey: PrefsKeys.KeyLastSyncFinishTime)
+            } else {
+                self.prefs.removeObjectForKey(PrefsKeys.KeyLastSyncFinishTime)
+            }
+        }
+    }
+
+    lazy var syncManagerAPI = RustSyncManagerAPI(logger: self.logger)
+
+    public func migrateSyncData() {
+        // The sync functions used by the old native sync manager did not return a
+        // persisted state string. And if we do not call the sync manager's rust
+        // function with persisted state data that call will be treated as the first
+        // sync for the account, which could take longer than the user expects to be
+        // completed. To avoid that, we sync one component with the rust sync manager
+        // and store the returned persisted state data for the first sync manager full
+        // sync.
+
+        var stateHasBeenSet = false
+
+        // XXX: `RustTogglableEngines` is not used here because we want to sync the
+        // engines in the specified order.
+        for engineName in ["tabs", "passwords", "bookmarks", "history"] {
+            if let enabled = self.prefsForSync.boolForKey("engine.\(engineName).enabled"),
+               enabled && !stateHasBeenSet {
+                syncRustEngines(why: .user, engines: [engineName]).upon({ [unowned self] result in
+                    if let persistedState = result.successValue?.persistedState {
+                        self.prefs.setString(persistedState,
+                                             forKey: PrefsKeys.RustSyncManagerPersistedState)
+                    }
+                })
+                stateHasBeenSet = true
+            }
+        }
+    }
+
+    public var isSyncing: Bool {
+        return syncDisplayState != nil && syncDisplayState! == .inProgress
+    }
+
+    public var syncDisplayState: SyncDisplayState?
+
+    // The dispatch queue for coordinating syncing and resetting the database.
+    fileprivate let syncQueue = DispatchQueue(label: "com.mozilla.firefox.sync")
+
+    @objc func syncOnTimer() {
+        self.syncEverything(why: .scheduled)
+        self.profile.pollCommands()
+    }
+
+    fileprivate func repeatingTimerAtInterval(
+        _ interval: TimeInterval,
+        selector: Selector
+    ) -> Timer {
+        return Timer.scheduledTimer(timeInterval: interval,
+                                    target: self,
+                                    selector: selector,
+                                    userInfo: nil,
+                                    repeats: true)
+    }
+
+    func syncEverythingSoon() {
+        self.doInBackgroundAfter(SyncConstants.SyncOnForegroundAfterMillis) {
```

Comment:
if I'm understanding correctly, the backgrounding of this is purely for the purpose of the delay correct? Like the actual syncing will dispatch back to the `syncQueue` (which in turn dispatches back to the `Dispatch.global` default queue?

Reply:
I only intend to use either `syncQueue` or `Dispatch.global` depending on NIsh's preference.